src/plugins/hidsim/gui/vmi_win_gui_parser.cpp

```diff
@@ -0,0 +1,1040 @@
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <wchar.h>
+#include <string.h>
+
+/* Offset retrieval */
+#include "vmi_win_gui_offsets.h"
+#include "vmi_win_gui_atoms.h"
+#include "vmi_win_gui_utils.h"
+#include "vmi_win_gui_filter_definitions.h"
+
+#include "vmi_win_gui_parser.h"
+
+struct rect* get_visible_rect_from_bitmask(
+    char* map, size_t n, int scanline, struct rect* r)
+{
+    struct rect* result = NULL;
+
+    size_t byte;
+    int bit_idx;
+    unsigned int bit;
+    int x0 = -1, x1 = -1, y0 = -1, y1 = -1;
+    bool is_y_hole = false;
+    bool is_x_hole = false;
+
+    for (int y = r->y0; y < r->y1; y++)
+    {
+        for (int x = r->x0; x < r->x1; x++)
+        {
+            byte = (y * scanline + x)/8;
+
+            /* Parts of a wnd can be outside of the desktop's frame */
+            if (byte >= n || x < 0)
+                continue;
+
+            bit_idx = x % 8;
+            bit = 0x80 >> bit_idx;
+            if (!(map[byte] & bit))
+            {
+                if (x0 == -1 && y0 == -1)
+                {
+                    x0 = x;
+                    y0 = y;
+                }
+
+                if (!is_x_hole)
+                    x1 = x;
+
+                if (!is_y_hole)
+                    y1 = y;
+            }
+            else
+            {
+                if (x == x0)
+                    is_y_hole = true;
+
+                if (y == y1)
+                    is_x_hole = true;
+            }
+        }
+    }
+    if (x1 != -1 && y1 != -1)
+    {
+        result = (struct rect*) malloc(sizeof(struct rect));
+        if (!result)
+        {
+            printf("[HIDSIM][MONITOR] Memory allocation for result-wnd failed\n");
+            return NULL;
+        }
+        result->x0 = x0;
+        result->x1 = x1;
+        result->y0 = y0;
+        result->y1 = y1;
+        result->w = x1 - x0;
+        result->h = y1 - y0;
+    }
+
+    return result;
+}
+
+void update_visibility_bitmask(char* map, size_t n, int scanline,
+    struct rect* r)
+{
+    size_t byte;
+    int bit_idx;
+    unsigned int bit;
+
+    for (int y = r->y0; y < r->y1; y++)
+    {
+        for (int x = r->x0; x < r->x1; x++)
+        {
+            byte = (y * scanline + x)/8;
+
+            if (x < 0 || byte >= n )
+                continue;
+
+            /* Parts of a wnd can be outside of the desktop's frame */
+            bit_idx = x % 8;
+            bit = 0x80 >> bit_idx;
+            map[byte] |= bit;
+        }
+    }
+}
+
+
+/*
+ * Determine, if windows is visible;
+ * important since invisible windows might have visible children
+ */
+bool is_wnd_visible(vmi_instance_t vmi, vmi_pid_t pid, addr_t wnd)
+{
+    uint32_t style = 0;
+
+    if (VMI_FAILURE == vmi_read_32_va(vmi, wnd + symbol_offsets.wnd_style, pid, (uint32_t*)&style))
```

Comment:
Redundant cast

Reply:
Changed in fe0998f12563d5015b04349d52677fb2e8a5a9cd